cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -101,41 +255,130 @@ private void refreshCredentials() throws IOException {
             .build();
 
     StsTokenExchangeResponse response = handler.exchangeToken();
-    this.accessBoundarySessionKey = response.getAccessBoundarySessionKey();
-    this.intermediateAccessToken = response.getAccessToken();
+    return new IntermediateCredentials(
+        getTokenFromResponse(response, sourceAccessToken), response.getAccessBoundarySessionKey());
+  }
 
-    // The STS endpoint will only return the expiration time for the intermediary token
+  /**
+   * Extracts the access token from the STS exchange response and sets the appropriate expiration
+   * time.
+   *
+   * @param response The STS token exchange response.
+   * @param sourceAccessToken The original access token used for the exchange.
+   * @return The intermediate access token.
+   */
+  private static AccessToken getTokenFromResponse(
+      StsTokenExchangeResponse response, AccessToken sourceAccessToken) {
+    AccessToken intermediateToken = response.getAccessToken();
+
+    // The STS endpoint will only return the expiration time for the intermediate token
     // if the original access token represents a service account.
-    // The intermediary token's expiration time will always match the source credential expiration.
+    // The intermediate token's expiration time will always match the source credential
+    // expiration.
     // When no expires_in is returned, we can copy the source credential's expiration time.
-    if (response.getAccessToken().getExpirationTime() == null) {
-      if (sourceAccessToken.getExpirationTime() != null) {
-        this.intermediateAccessToken =
-            new AccessToken(
-                response.getAccessToken().getTokenValue(), sourceAccessToken.getExpirationTime());
-      }
+    if (intermediateToken.getExpirationTime() == null
+        && sourceAccessToken.getExpirationTime() != null) {
+      return new AccessToken(
+          intermediateToken.getTokenValue(), sourceAccessToken.getExpirationTime());
     }
+    return intermediateToken; // Return original if no modification needed
   }
 
-  private void refreshCredentialsIfRequired() {
-    // TODO(negarb): Implement refreshCredentialsIfRequired
-    throw new UnsupportedOperationException("refreshCredentialsIfRequired is not yet implemented.");
+  /**
+   * Completes the refresh task by storing the results and clearing the single flight slot.
+   *
+   * <p>This method is called when a refresh task finishes. It stores the refreshed credentials if
+   * successful. The single-flight "slot" is cleared, allowing subsequent refresh attempts. Any
+   * exceptions during the refresh are caught and suppressed to prevent indefinite blocking of
+   * subsequent refresh attempts.
+   */
+  private void finishRefreshTask(ListenableFuture<IntermediateCredentials> finishedTask) {
+    synchronized (refreshLock) {
+      try {
+        this.intermediateCredentials = Futures.getDone(finishedTask);
+      } catch (Exception e) {
+        // noop
```

Comment:
Good catch! OAuth2Credentials had implemented this function like this, I think maybe to avoid throwing exceptions in the async mode? I'm not sure! I did some testing and the way it's implemented right now causes the blocking refresh to fail silently without notifying anywhere that it has failed. I'm adding some error handling and will send a new commit soon.

Reply:
Fixed! Now the blocking refresh will correctly propagate the exception if something goes wrong during the refresh.